#167512374 User can get article category subscribed for

Jonathan4github opened the pull request:
#### What does this PR do?
This Pull request implements user should be able to get articles category subscribed for
#### Description of Task to be completed
- create a subscription model
- implement functionality in the controller file
#### How should this be manually tested?
 - clone this repository
- ```cd``` into the repo and run ```npm install```
- using postman **SET**  ```Authorization``` to ```Bearer <user token>```
- make a **GET**  request to /api/v1/subscriptions 
#### Any Background Context You want to provide?

#### What are the relevant pivotal tracker stories?

[#167512374](https://pivotaltracker.com/story/show/167512374)

#### Screenshots (if appropriate)
<img width="1126" alt="Screenshot 2019-08-19 at 9 07 23 PM" src="https://user-images.githubusercontent.com/31247452/63295803-7b085400-c2c5-11e9-8727-5d39d8dd368f.png">

#### Questions:

ibkadeeko committed:
feat(articles): upvote and downvote articles (#31)

- implement upvote and downvote functionality for articles
- only logged in users can vote
- users can upvote or downvote an article only once
- users can change their previous vote
- users can retract their vote
- article is suspended when difference between down votes and upvotes is >= 5

[Delivers #167377547]

Jonathan4github commented on tests/routes/articles.test.js:
```diff
@@ -883,3 +895,48 @@ describe('GET /articles/articleId/inline_comments', () => {
+    status.should.eql(404);
+  });
+
+  it('should return 200', async () => {
```
Add message for 404

Jonathan4github commented on src/controllers/article.js:
```diff
@@ -516,5 +521,62 @@ export default {
+   *
+   * @return {Object} - callback that execute the controller
+   */
+  getArticlesSubscribed: async (request, response) => {
```
getSubscribedArticles

Jonathan4github commented on src/controllers/article.js:
```diff
@@ -516,5 +521,62 @@ export default {
+    const { id: userId } = request.user;
+    const { page = 1, limit = 10 } = request.query;
+
+    const user = await Users.findByPk(userId);
```
Move to verify token middleware.

Jonathan4github commented on src/controllers/article.js:
```diff
@@ -516,5 +521,62 @@ export default {
+
+    if (!user) throw new NotFoundError();
+
+    const categories = (await user.getSubscriptions({ attributes: ['categoryId'] })).map(sub => sub.categoryId);
```
categoryIds

iverenshaguy replied:
Why wasn't this changed?

Jonathan4github commented on src/controllers/article.js:
```diff
@@ -516,5 +521,62 @@ export default {
+      }]
+    };
+
+    const { data, ...otherDatas } = await paginator(Articles, query);
```
other data should be singular

Jonathan4github commented on src/controllers/article.js:
```diff
@@ -516,5 +521,62 @@ export default {
+      status: 'success',
+      data: articles,
+      ...otherDatas,
+      message: 'Article category subscribed for retrieved successfully'
```
```suggestion
      message: 'Article categories subscribed for retrieved successfully'
```

Jonathan4github commented on src/routes/article.js:
```diff
@@ -296,6 +297,8 @@ router.post(
   asyncWrapper(createArticle)
 );
 
+router.get('/subscriptions', asyncWrapper(verifyToken), asyncWrapper(getArticlesSubscribed));
```
Swagger doc

Jonathan4github commented on tests/routes/articles.test.js:
```diff
@@ -883,3 +895,48 @@ describe('GET /articles/articleId/inline_comments', () => {
+    await Articles.destroy({ where: {} });
+    await ArticleCategories.destroy({ where: {} });
+    await Subscriptions.destroy({ where: {} });
+  });
```
newline and make check what test is to return.

iverenshaguy commented on tests/routes/profile.test.js:
```diff
@@ -30,11 +30,14 @@ let currentUserId;
       .post(signUrl)
       .send(usersignUpdetail);
   });
+  after(async () => {
```
Leave a new line above and below this `after` block

iverenshaguy commented on tests/routes/articles.test.js:
```diff
@@ -89,13 +96,20 @@ describe('Articles API', () => {
       const response = await chai.request(app)
         .get(`${baseRoute}/articles/feed`)
-        .set('Authorization', `Bearer ${userToken}`);
+        .set('Authorization', `Bearer ${randomUserToken}`);
```
Why was this changed? Why should a user be random?

Jonathan4github replied:
When I added a check if the user exists in the authentication middleware. Most tests failed because the "userToken" does not exist in the database so I changed them to randomUserToken.

iverenshaguy commented on src/middlewares/authentication.js:
```diff
@@ -33,11 +35,19 @@ export default {
 
-      next();
+      const { id: userId } = request.user;
+
```
Please remove this new line

iverenshaguy commented on src/middlewares/authentication.js:
```diff
@@ -33,11 +35,19 @@ export default {
-    jwt.verify(token, SECRET_KEY, (error, decodedToken) => {
-      if (error) throw new ApplicationError(401, `${error.message}`);
+    jwt.verify(token, SECRET_KEY, async (error, decodedToken) => {
+      if (error) return next(new ApplicationError(401, `${error.message}`));
```
Why was `throw` changed to return? What benefit does this give us? I know express automatically uses the error middleware when `next()` is called with an argument, what benefit does this give us over simply throwing the error which does the same thing?

iverenshaguy commented on src/middlewares/authentication.js:
```diff
@@ -33,11 +35,19 @@ export default {
+
+      const user = await Users.findByPk(userId);
+
+      if (!user) return next(new ApplicationError(403, 'User not found'));
```
Must we return next? Why not just throw it? Any benefit?

iverenshaguy reviewed (changes requested):
Please review my feedback

iverenshaguy committed:
feat(subscriptions): implement get article category subscribed for

- implement get article category subscribed functionality
- add pagination to articles

[Finishes #167512374]
